servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultAsyncContextProvider.java

```diff
@@ -333,6 +313,31 @@ public <T, U, V> BiFunction<T, U, V> wrapBiFunction(final BiFunction<T, U, V> fu
         return new ContextPreservingBiFunction<>(func, context);
     }
 
+    private static ContextMap exchangeContext(ContextMap contextMap) {
+        ContextMap result;
+        final Thread currentThread = Thread.currentThread();
+        if (currentThread instanceof ContextMapHolder) {
+            final ContextMapHolder asyncContextMapHolder = (ContextMapHolder) currentThread;
+            result = asyncContextMapHolder.context();
+            if (result == null) {
+                result = newContextMap();
+            }
+            asyncContextMapHolder.context(contextMap);
+        } else {
+            result = CONTEXT_THREAD_LOCAL.get();
+            CONTEXT_THREAD_LOCAL.set(contextMap);
+        }
+        return result;
+    }
+
+    private static void detachContext(ContextMap expectedContext, ContextMap toRestore) {
+        ContextMap current = exchangeContext(toRestore);
+        if (current != expectedContext) {
+            LOGGER.debug("Current context didn't match the expected context. current: {}, expected: {}",
+                    current, expectedContext);
```

Comment:
For the follow-up: would be great to also add `new Throwable()` at the end to help us track where this comes from. To avoid unnecessary stack-trace capture, consider using `NOT_IS_DEBUG_ENABLED`.

Reply:
Done in https://github.com/apple/servicetalk/pull/3183/commits/edf142afc86415dbcee94025947c7d40074a262a